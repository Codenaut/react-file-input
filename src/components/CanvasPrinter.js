import React, { PureComponent } from 'react';

import '../styles/CanvasPrinter.scss';

class CanvasPrinter extends PureComponent {
  constructor(props) {
    super(props);

    this.canvas = null;

    this.printImage = this.printImage.bind(this);
    this.cropAndResizeImage = this.cropAndResizeImage.bind(this);
    this.resizeImage = this.resizeImage.bind(this);
  }

  /**
   * [printImage description]
   * @param  {Image} image
   * @param  {{x: Number, y: Number}} offset
   * @param  {{width: Number, height: Number}} sourceSize
   * @param  {{width: Number, height: Number}} outputSize
   * @return {Promise<Blob>} Output image blob
   * @see https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/drawImage
   */
  printImage(image, offset, sourceSize, outputSize) {
    this.canvas.width = outputSize.width;
    this.canvas.height = outputSize.height;

    this.canvas
      .getContext('2d')
      .drawImage(
        image,
        offset.x,
        offset.y,
        sourceSize.width,
        sourceSize.height,
        0,
        0,
        outputSize.width,
        outputSize.height
      );

    return new Promise(resolve => {
      this.canvas.toBlob(resolve);
    });
  }

  cropAndResizeImage(image, area, scaleOptions = null) {
    const { clientHeight, clientWidth, naturalHeight, naturalWidth } = image;
    const { x0, y0, x1, y1 } = area;

    const widthFactor = naturalWidth / clientWidth;
    const heightFactor = naturalHeight / clientHeight;

    const offset = {
      x: x0 * widthFactor,
      y: y0 * heightFactor,
    };

    const sourceSize = {
      width: (x1 - x0 ) * widthFactor,
      height: (y1 - y0 ) * heightFactor,
    };

    return this.printImage(image, offset, sourceSize, scaleOptions || sourceSize);
  }

  resizeImage(image, scaleOptions) {
    if (!scaleOptions.keepAspectRatio) {
      return this.printImage(
        image,
        { x: 0, y: 0 },
        { width: image.naturalWidth, height: image.naturalHeight },
        scaleOptions
      );
    }

    const { naturalWidth, naturalHeight } = image;

    const aspectRatio = scaleOptions.width / scaleOptions.height;
    const horizontal = naturalWidth < naturalHeight * aspectRatio;

    const offset = !horizontal
      ? { x: (naturalWidth - (naturalHeight * aspectRatio)) / 2, y: 0 }
      : { x: 0, y: (naturalHeight - (naturalWidth / aspectRatio)) / 2 };

    const sourceSize = { width: naturalWidth - (2 * offset.x), height: naturalHeight - (2 * offset.y) };

    return this.printImage(image, offset, sourceSize, scaleOptions);
  }

  render() {
    return (
      <canvas
        className="brainhub-canvas-printer"
        ref={ref => {
          this.canvas = ref;
        }}
      />
    );
  }
}

export default CanvasPrinter;
