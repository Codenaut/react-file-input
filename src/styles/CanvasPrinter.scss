.brainhub-canvas-printer {
  display: none;
}
