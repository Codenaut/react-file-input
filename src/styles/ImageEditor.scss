@import 'global';

$POINT_SIZE: 8px;

.brainhub-crop-tool,
.brainhub-image-rotator {
  position: relative;
  overflow: hidden;
}

.brainhub-crop-tool {
  cursor: crosshair;
}

.brainhub-crop-tool__controls,
.brainhub-image-rotator__controls {
  text-align: center;
}

.brainhub-crop-tool__image,
.brainhub-image-rotator__image {
  max-width: 100%;
  max-height: 100%;
  display: block;
  margin: 0 auto;
}

.brainhub-crop-tool__button,
.brainhub-image-rotator__button {
  border: none;
  outline: none;
  font-size: 14px;
  color: white;
  border-radius: 5px;
  padding: 4px 16px;
  margin: 4px;
}

.brainhub-crop-tool__button--cancel,
.brainhub-image-rotator__button--arrow {
  background-color: $LIGTH_GRAY;
}

.brainhub-crop-tool__button--save,
.brainhub-image-rotator__button--save {
  background-color: $LIGHT_GREEN;
}

.brainhub-selected-area {
  position: absolute;
  border: 1px dashed white;
  cursor: move;
  box-shadow: 0 0 0 9999em rgba(0, 0, 0, 0.5);
}

.brainhub-selected-area__resize-point {
  width:8px;
  height: 8px;
  position: absolute;
  background-color: white;
  border: 1px solid;
}

.brainhub-selected-area__resize-point--t,
.brainhub-selected-area__resize-point--b {
  left: calc(50% - 6px); // border + point_size/2
}

.brainhub-selected-area__resize-point--r,
.brainhub-selected-area__resize-point--l {
  top: calc(50% - 6px); // border + point_size/2
}

.brainhub-selected-area__resize-point--tl,
.brainhub-selected-area__resize-point--t,
.brainhub-selected-area__resize-point--tr {
  top: calc(0px - 6px); // border + point_size/2
}
.brainhub-selected-area__resize-point--tr,
.brainhub-selected-area__resize-point--r,
.brainhub-selected-area__resize-point--br {
  right: calc(0px - 6px); // border + point_size/2
}
.brainhub-selected-area__resize-point--bl,
.brainhub-selected-area__resize-point--b,
.brainhub-selected-area__resize-point--br {
  bottom: calc(0px - 6px); // border + point_size/2
}
.brainhub-selected-area__resize-point--tl,
.brainhub-selected-area__resize-point--l,
.brainhub-selected-area__resize-point--bl {
  left: calc(0px - 6px); // border + point_size/2
}

.brainhub-selected-area__resize-point--t,
.brainhub-selected-area__resize-point--b {
  cursor: ns-resize;
}
.brainhub-selected-area__resize-point--tr,
.brainhub-selected-area__resize-point--bl {
  cursor: nesw-resize;
}
.brainhub-selected-area__resize-point--l,
.brainhub-selected-area__resize-point--r {
  cursor: ew-resize;
}
.brainhub-selected-area__resize-point--tl,
.brainhub-selected-area__resize-point--br {
  cursor: nwse-resize;
}

.brainhub-canvas-printer {
  display: none;
}
