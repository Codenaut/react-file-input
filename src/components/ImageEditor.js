import React, { Component } from 'react';
import PropTypes from 'prop-types';

import SelectedArea from './SelectedArea';

import '../styles/ImageEditor.scss';

const between = (value, max, min =0) => Math.max(Math.min(value, max), min);

/**
 * @todo Change selected area for selection
 */

class ImageEditor extends Component {
  constructor(props) {
    super(props);

    this.state = {
      landscape: true,
      isResizing: false,
      resizeLandscape: false,
      x0: 0,
      y0: 0,
      x1: 0,
      y1: 0,
      isMoving: false,
      xM: 0,
      yM: 0,
    };

    this.container = null;
    this.image = null;

    this.onImageLoad = this.onImageLoad.bind(this);
    this.getSelectedAreaBoundaries = this.getSelectedAreaBoundaries.bind(this);

    this.onMouseMove = this.onMouseMove.bind(this);
    this.onMouseUp = this.onMouseUp.bind(this);

    this.startResize = this.startResize.bind(this);
    this.resize = this.resize.bind(this);
    this.resizeSide = this.resizeSide.bind(this);

    this.startMove = this.startMove.bind(this);
    this.move = this.move.bind(this);
  }

  componentWillMount() {
    document.addEventListener('mousemove', this.onMouseMove, false);
    document.addEventListener('onmousemove', this.onMouseMove, false);
    document.addEventListener('mouseup', this.onMouseUp, false);
  }

  componentWillUnmount() {
    document.removeEventListener('mousemove', this.onMouseMove);
    document.removeEventListener('mouseup', this.onMouseUp);
  }

  /**
   * Set initial selection area according to the maximum posible, and the landscape orientation
   */
  onImageLoad() {
    const { ratio } = this.props;
    const { clientWidth: width, clientHeight: height } = this.image;

    const landscape = width > height * ratio;

    const selectionWidth = landscape ? (height * ratio) : width;
    const selectionHeight = landscape ? height : (width / ratio);

    const xOffset = landscape ? ((width - selectionWidth) / 2) : 0;
    const yOffset = landscape ? 0 : ((height - selectionHeight) / 2);

    this.setState(state => ({
      ...state,
      landscape,
      x0: xOffset,
      y0: yOffset,
      x1: width - xOffset,
      y1: height - yOffset,
    }));
  }

  startMove(event) {
    event.preventDefault();
    const { pageX: mouseX, pageY: mouseY } = event;
    const { clientWidth: width, clientHeight: height, offsetLeft: imageLeft, offsetTop: imageTop } = this.image;
    const { offsetLeft: containerLeft, offsetTop: containerTop } = this.container;

    this.setState({
      isMoving: true,
      xM: between(mouseX - imageLeft - containerLeft, width),
      yM: between(mouseY - imageTop - containerTop, height),
    });
  }

  move(event) {
    event.preventDefault();
    const { x0, y0, x1, y1, xM, yM } = this.state;
    const { pageX: mouseX, pageY: mouseY } = event;
    const { clientWidth: width, clientHeight: height, offsetLeft: imageLeft, offsetTop: imageTop } = this.image;
    const { offsetLeft: containerLeft, offsetTop: containerTop } = this.container;

    // Calculates the real posible point (i.e point tha falls between image boundaries)

    const clickX = between(mouseX - containerLeft - imageLeft, width);
    const clickY = between(mouseY - containerTop - imageTop, height);

    const dX = clickX - xM;
    const dY = clickY - yM;

    // processX
    const realDX = dX > 0
      ? Math.min(dX, width - x1)
      : Math.max(dX, -x0);

    const realDY = dY > 0
      ? Math.min(dY, height - y1)
      : Math.max(dY, -y0);

    this.setState(state => ({
      xM: clickX,
      yM: clickY,
      x0: x0 + realDX,
      y0: y0 + realDY,
      x1: x1 + realDX,
      y1: y1 + realDY,
    }));
  }

  /**
   * Takes the selection points relative to the image and transform it into css absolute properties relative to the wrapping div
   * @return {Object} Object containing top, right, bottom and left css properties in percentage (or empty)
   */
  getSelectedAreaBoundaries() {
    if (!this.container) {
      return {};
    }
    const { x0, y0, x1, y1 } = this.state;
    const { clientWidth: width, clientHeight: height } = this.container;
    const { offsetLeft, offsetTop } = this.image;

    return {
      top: `${100*(Math.min(y0, y1) + offsetTop) / height}%`,
      right: `${100*(width - Math.max(x0, x1) - offsetLeft) / width}%`,
      bottom: `${100*(height - Math.max(y0, y1) - offsetTop) / height}%`,
      left: `${100*(Math.min(x0, x1) + offsetLeft) / width}%`,
    };
  }

  startResize(event, resizeLandscape, invertX = false, invertY = false, isResizingSide = false) {
    event.preventDefault();

    this.setState(state => ({
      ...state,
      isResizing: !isResizingSide,
      isResizingSide,
      resizeLandscape: !resizeLandscape,
      x0: invertX ? state.x1 : state.x0,
      y0: invertY ? state.y1 : state.y0,
      x1: invertX ? state.x0: state.x1,
      y1: invertY ? state.y0 : state.y1,
    }));
  }

  resize(event) {
    event.preventDefault();

    const { ratio } = this.props;
    const { x0, y0, x1, y1, resizeLandscape: landscape } = this.state;
    const { clientWidth: width, clientHeight: height, offsetLeft: imageLeft, offsetTop: imageTop } = this.image;
    const { offsetLeft: containerLeft, offsetTop: containerTop } = this.container;

    const { pageX: mouseX, pageY: mouseY } = event;

    // Calculate points relative to the image
    const clickX = mouseX - containerLeft - imageLeft;
    const clickY = mouseY - containerTop - imageTop;

    // Calculates the real posible point (i.e point tha falls between image boundaries)
    const x = between(clickX, width);
    const y = between(clickY, height);

    const resizeFactor = ((x0 > x1 && y0 < y1) || (y0 > y1 && x0 < x1)) ? -1 : 1;

    const point1 = landscape
      ? [resizeFactor * (y - y0) * ratio + x0, y]
      : [x, resizeFactor * (x - x0) / ratio + y0];

    if (point1[0] > width || point1[1] > height) {
      point1[0] = landscape ? width : ( resizeFactor * (height - y0) * ratio + x0);
      point1[1] = landscape ? (resizeFactor * (width - x0) / ratio + y0) : height;
    } else if (point1[0] < 0 || point1[1] < 0) {
      point1[0] = landscape ? 0 : ( - resizeFactor * (y0 * ratio) + x0);
      point1[1] = landscape ? ( - resizeFactor * (x0 / ratio) + y0) : 0;
    }

    this.setState({ x1: point1[0], y1: point1[1] });
  }

  resizeSide(event) {
    event.preventDefault();

    const { ratio } = this.props;
    const { x0, y0, x1, y1, resizeLandscape: landscape } = this.state; // landscap aka vertical rules
    const { clientWidth: width, clientHeight: height, offsetLeft: imageLeft, offsetTop: imageTop } = this.image;
    const { offsetLeft: containerLeft, offsetTop: containerTop } = this.container;

    const { pageX: mouseX, pageY: mouseY } = event;

    // Calculate points relative to the image
    const clickX = mouseX - containerLeft - imageLeft;
    const clickY = mouseY - containerTop - imageTop;

    // Calculates the real posible point (i.e point tha falls between image boundaries)
    const x = between(clickX, width);
    const y = between(clickY, height);

    const resizeFactor = ((x0 > x1 && y0 < y1) || (y0 > y1 && x0 < x1)) ? -1 : 1;

    const point1 = landscape
      ? [[- resizeFactor * (y -y1) * ratio /2 + x0, y0], [resizeFactor * (y - y1) * ratio/2 + x1, y]]
      : [[x0, - resizeFactor * (x - x1) / 2 / ratio + y0], [x, resizeFactor * (x - x1) / 2 / ratio + y1]];

    if (point1[0][1] > height || point1[1][1] < 0
      || point1[0][1] < 0 || point1[1][1] > height
      || point1[0][0] > width || point1[1][0] < 0
      || point1[0][0] < 0 || point1[1][0] > width) {
      const adjustFactor = landscape ? Math.sign(x1-x0) : Math.sign(y1 - y0);

      const maxDY = Math.max(Math.min(height - y0, y1, y0, height - y1), 0);
      const maxDX = Math.max(Math.min(width - x0, x1, x0, width - x1), 0);

      point1[0][0] = landscape
        ? x0 - adjustFactor * maxDX
        :point1[0][0];
      point1[0][1] = landscape
        ? point1[0][1]
        :y0 - adjustFactor * maxDY;
      point1[1][0] = landscape
        ? x1 + adjustFactor * maxDX
        :x1 + adjustFactor * resizeFactor * maxDY * 2 * ratio;
      point1[1][1] = landscape
        ? y1 + adjustFactor * resizeFactor * maxDX * 2 /ratio
        : y1 + adjustFactor * maxDY;
    }

    this.setState(state => ({ ...state, x0: point1[0][0], y0: point1[0][1], x1: point1[1][0], y1: point1[1][1] }));
  }

  onMouseMove(event) {
    const { isResizing, isResizingSide, isMoving }= this.state;
    if (isMoving) {
      this.move(event);
    } else if (isResizing) {
      this.resize(event);
    } else if (isResizingSide) {
      this.resizeSide(event);
    }
  }

  onMouseUp(event) {
    const { isResizing, isResizingSide, isMoving } = this.state;

    if ( isResizing || isResizingSide || isMoving) {
      event.preventDefault();

      this.setState(state => ({
        ...state,
        isResizing: false,
        isResizingSide: false,
        x0: Math.min(state.x0, state.x1),
        y0: Math.min(state.y0, state.y1),
        x1: Math.max(state.x0, state.x1),
        y1: Math.max(state.y0, state.y1),
        isMoving: false,
        xM: 0,
        yM: 0,
      }));
    }
  }

  render() {
    const { image } = this.props;

    return (
      <div
        className="brainhub-image-editor"
        ref={ref => {
          this.container = ref;
        }}
        onMouseDown={this.onMouseDown}
        onDragStart={event => {
          event.preventDefault();
        }}
      >
        <img
          ref={ref => {
            this.image = ref;
          }}
          className="brainhub-image-editor__image"
          src={image}
          onLoad={this.onImageLoad}/>
        <SelectedArea
          startMove={this.startMove}
          startResize={this.startResize}
          style={this.getSelectedAreaBoundaries()}
        />
      </div>);
  }
}

ImageEditor.defaultProps = {
  ratio: 2/3, // w/h
};

ImageEditor.propTypes = {
  image: PropTypes.string.isRequired,
  ratio: PropTypes.number,
};

export default ImageEditor;
