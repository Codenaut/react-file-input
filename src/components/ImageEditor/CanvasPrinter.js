import React, { PureComponent } from 'react';

import '../../styles/ImageEditor.scss';

const calculateOffset = (width, height, scaleOptions, inverted = false) => {
  const aspectRatio = inverted
    ? scaleOptions.height / scaleOptions.width
    : scaleOptions.width / scaleOptions.height;

  const horizontal = width < height * aspectRatio;

  return !horizontal
    ? { x: (width - (height * aspectRatio)) / 2, y: 0 }
    : { x: 0, y: (height - (width / aspectRatio)) / 2 };
};


class CanvasPrinter extends PureComponent {
  constructor(props) {
    super(props);

    this.canvas = null;

    this.printImage = this.printImage.bind(this);
    this.cropAndResizeImage = this.cropAndResizeImage.bind(this);
    this.resizeImage = this.resizeImage.bind(this);
  }

  /**
   * [printImage description]
   * @param  {Image} image
   * @param  {{x: Number, y: Number}} offset
   * @param  {{width: Number, height: Number}} sourceSize
   * @param  {{width: Number, height: Number}} outputSize
   * @return {Promise<Blob>} Output image blob
   * @see https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/drawImage
   */
  printImage(image, offset, sourceSize, outputSize) {
    this.canvas.width = outputSize.width;
    this.canvas.height = outputSize.height;

    this.canvas
      .getContext('2d')
      .drawImage(
        image,
        offset.x,
        offset.y,
        sourceSize.width,
        sourceSize.height,
        0,
        0,
        outputSize.width,
        outputSize.height
      );

    return new Promise(resolve => {
      this.canvas.toBlob(resolve);
    });
  }

  cropAndResizeImage(image, area, scaleOptions = null) {
    const { clientHeight, clientWidth, naturalHeight, naturalWidth } = image;
    const { x0, y0, x1, y1 } = area;

    const widthFactor = naturalWidth / clientWidth;
    const heightFactor = naturalHeight / clientHeight;

    const offset = {
      x: x0 * widthFactor,
      y: y0 * heightFactor,
    };

    const sourceSize = {
      width: (x1 - x0 ) * widthFactor,
      height: (y1 - y0 ) * heightFactor,
    };

    return this.printImage(image, offset, sourceSize, scaleOptions || sourceSize);
  }

  resizeImage(image, scaleOptions) {
    if (!scaleOptions.keepAspectRatio) {
      return this.printImage(
        image,
        { x: 0, y: 0 },
        { width: image.naturalWidth, height: image.naturalHeight },
        scaleOptions
      );
    }

    const { naturalWidth, naturalHeight } = image;

    const aspectRatio = scaleOptions.width / scaleOptions.height;
    const horizontal = naturalWidth < naturalHeight * aspectRatio;

    const offset = !horizontal
      ? { x: (naturalWidth - (naturalHeight * aspectRatio)) / 2, y: 0 }
      : { x: 0, y: (naturalHeight - (naturalWidth / aspectRatio)) / 2 };

    const sourceSize = { width: naturalWidth - (2 * offset.x), height: naturalWidth - (2 * offset.x) };

    return this.printImage(image, offset, sourceSize, scaleOptions);
  }


  rotateAndScaleImage(image, angle, scaleOptions = null) {
    const { naturalWidth, naturalHeight } = image;
    const baseWidth = (scaleOptions && scaleOptions.width) || naturalWidth;
    const baseHeight = (scaleOptions && scaleOptions.height) || naturalHeight;

    const inverted = !(angle %2 === 0);

    const width = (inverted && baseHeight) || baseWidth;
    const height = (inverted && baseWidth) || baseHeight;

    this.canvas.width = baseWidth;
    this.canvas.height = baseHeight;

    const offset = scaleOptions && scaleOptions.keepAspectRatio
      ? calculateOffset(naturalWidth, naturalHeight, scaleOptions, inverted)
      : { x: 0, y: 0 };

    const context = this.canvas.getContext('2d');

    context.translate(baseWidth / 2, baseHeight / 2);
    context.rotate((angle * Math.PI) /2);
    context.drawImage(
      image,
      offset.x,
      offset.y,
      naturalWidth - (2 * offset.x),
      naturalHeight - (2 * offset.y),
      - width / 2,
      - height / 2,
      width,
      height,
    );

    return new Promise(resolve => {
      this.canvas.toBlob(resolve);
    });
  }

  rotateImage(image, angle, scaleOptions = null) {
    const { naturalWidth, naturalHeight } = image;

    const inverted = !(angle %2 === 0);

    const width = (inverted && naturalHeight) || naturalWidth;
    const height = (inverted && naturalWidth) || naturalHeight;

    this.canvas.width = width;
    this.canvas.height = height;

    const context = this.canvas.getContext('2d');

    context.translate(width / 2, height / 2);
    context.rotate((angle * Math.PI) /2);
    context.drawImage(
      image,
      - naturalWidth / 2,
      - naturalHeight / 2,
    );

    return new Promise(resolve => {
      this.canvas.toBlob(resolve);
    });
  }

  render() {
    return (
      <canvas
        className="brainhub-canvas-printer"
        ref={ref => {
          this.canvas = ref;
        }}
      />
    );
  }
}

export default CanvasPrinter;
