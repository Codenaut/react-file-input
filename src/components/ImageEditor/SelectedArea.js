import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import '../../styles/ImageEditor.scss';

class SelectedArea extends PureComponent {
  constructor(props) {
    super(props);

    this.selectedArea = null;

    this.startMove = this.startMove.bind(this);

    this.resizeTopLeft = this.resizeTopLeft.bind(this);
    this.resizeTop = this.resizeTop.bind(this);
    this.resizeTopRight = this.resizeTopRight.bind(this);
    this.resizeRight = this.resizeRight.bind(this);
    this.resizeBottomRight = this.resizeBottomRight.bind(this);
    this.resizeBottom = this.resizeBottom.bind(this);
    this.resizeBottomLeft = this.resizeBottomLeft.bind(this);
    this.resizeLeft = this.resizeLeft.bind(this);
  }

  resizeTopLeft(event) {
    const { startResize } = this.props;

    startResize(event, true, true, false);
  }

  resizeTop(event) {
    const { startResize } = this.props;

    startResize(event, false, true, true, true);
  }

  resizeTopRight(event) {
    const { startResize } = this.props;

    startResize(event, false, true);
  }

  resizeRight(event) {
    const { startResize } = this.props;

    startResize(event, false, false, true);
  }

  resizeBottomRight(event) {
    const { startResize } = this.props;

    startResize(event, false, false);
  }

  resizeBottom(event) {
    const { startResize } = this.props;

    startResize(event, false, false, true, true);
  }

  resizeBottomLeft(event) {
    const { startResize } = this.props;

    startResize(event, true, false);
  }

  resizeLeft(event) {
    const { startResize } = this.props;

    startResize(event, true, false, true);
  }

  startMove(event) {
    if (event.target === this.selectedArea) {
      const { startMove } = this.props;

      startMove(event);
    }
  }

  render() {
    const { style } = this.props;

    return (
      <div
        className="brainhub-image-editor__selected-area"
        ref={ref => {
          this.selectedArea = ref;
        }}
        style={style}
        onMouseDown={this.startMove} >
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--tl"
          onMouseDown={this.resizeTopLeft}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--t"
          onMouseDown={this.resizeTop}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--tr"
          onMouseDown={this.resizeTopRight}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--r"
          onMouseDown={this.resizeRight}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--br"
          onMouseDown={this.resizeBottomRight}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--b"
          onMouseDown={this.resizeBottom}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--bl"
          onMouseDown={this.resizeBottomLeft}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--l"
          onMouseDown={this.resizeLeft}
        />
      </div>
    );
  }
}

SelectedArea.defaultProps = {
  style: {
    top: '0',
    right: '0',
    bottom: '0',
    left: '0',
  },
};

SelectedArea.propTypes = {
  style: PropTypes.shape({
    top: PropTypes.string,
    right: PropTypes.string,
    bottom: PropTypes.string,
    left: PropTypes.string,
  }),
  startMove: PropTypes.func.isRequired,
  startResize: PropTypes.func.isRequired,
};

export default SelectedArea;
