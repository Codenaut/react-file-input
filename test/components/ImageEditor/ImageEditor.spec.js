import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import ImageEditor from '../../../src/components/ImageEditor';
import SelectedArea from '../../../src/components/ImageEditor/SelectedArea';
import CanvasPrinter from '../../../src/components/ImageEditor/CanvasPrinter';

const defaultProps = {
  image: '',
  onCancelEdition: () => null,
  onEdition: () => null,
};

const setup = (props = {}) => {
  const imageEditor = mount(<ImageEditor {...defaultProps} {...props} />);

  return {
    imageEditor,
    selectedArea: imageEditor.find(SelectedArea),
    canvasPrinter: imageEditor.find(CanvasPrinter),
    button: imageEditor.find('button'),
  };
};

describe('components', () => {
  describe('ImageEditor', () => {
    it('should render with a default state', () => {
      const { imageEditor } = setup();
      expect(imageEditor.state()).toEqual({
        landscape: true,
        isMoving: false,
        isResizing: false,
        resizeVertical: false,
        x0: 0,
        y0: 0,
        x1: 0,
        y1: 0,
        xM: 0,
        yM: 0,
      });
    });

    it('should render a SelectedArea', () => {
      const { selectedArea } = setup();

      expect(selectedArea).toHaveLength(1);
    });

    it('should render a CanvasPrinter', () => {
      const { canvasPrinter } = setup();

      expect(canvasPrinter).toHaveLength(1);
    });

    it('should render two buttons', () => {
      const { button } = setup();

      expect(button).toHaveLength(2);
      expect(button.first().text()).toBe('Cancel');
      expect(button.last().text()).toBe('Save image');
    });

    it('should cancel render when Cancel button is clicked', () => {
      const onCancelEdition = jest.fn();
      const { button } = setup({ onCancelEdition });

      const cancelButton = button.first();

      cancelButton.simulate('click');

      expect(onCancelEdition).toHaveBeenCalled();
    });

    it('should start moving when startMove is called', () => {
      const { imageEditor } = setup();
      const event = {
        preventDefault: () => null,
      };
      imageEditor.instance().startMove(event);
      expect(imageEditor.state('isMoving')).toBeTruthy();
    });

    it('should only interact to mouse movement is the state isMoving or isResizing', () => {
      const { imageEditor } = setup();
      const event = {
        preventDefault: () => null,
      };
      const imageEditorInstance = imageEditor.instance();
      const handleMoveSpy = jest.spyOn(imageEditorInstance, 'handleMove');
      const handleReiszeSpy = jest.spyOn(imageEditorInstance, 'handleResize');

      imageEditorInstance.onMouseMove(event);

      expect(handleMoveSpy).not.toHaveBeenCalled();
      expect(handleReiszeSpy).not.toHaveBeenCalled();

      imageEditor.setState({ isMoving: true });
      imageEditorInstance.onMouseMove(event);

      expect(handleMoveSpy).toHaveBeenCalledTimes(1);
      expect(handleReiszeSpy).not.toHaveBeenCalled();

      imageEditor.setState({ isMoving: false, isResizing: true });
      imageEditorInstance.onMouseMove(event);

      expect(handleMoveSpy).toHaveBeenCalledTimes(1);
      expect(handleReiszeSpy).toHaveBeenCalledTimes(1);
    });

    it('should unset isMoving and isResizing and order the selected area points when mouseup', () => {
      const { imageEditor } = setup();

      const prevState = {
        landscape: true,
        isMoving: false,
        isResizing: true,
        isResizingSide: true,
        resizeVertical: true,
        x0: 50,
        y0: 50,
        x1: 10,
        y1: 10,
        xM: 10,
        yM: 10,
      };
      imageEditor.setState(prevState);

      const event = {
        preventDefault: () => null,
      };

      imageEditor.instance().onMouseUp(event);

      expect(imageEditor.state()).toEqual({
        ...prevState,
        isMoving: false,
        isResizing: false,
        isResizingSide: false,
        resizeVertical: false,
        x0: 10,
        y0: 10,
        x1: 50,
        y1: 50,
        xM: 0,
        yM: 0,
      });
    });

    it('should match exact snapshot', () => {
      const tree = renderer.create(
        <div>
          <ImageEditor {...defaultProps} />
        </div>
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
