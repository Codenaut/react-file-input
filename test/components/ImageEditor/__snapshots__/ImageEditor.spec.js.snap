// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components ImageEditor should match exact snapshot 1`] = `
<div>
  <div>
    <div>
      <div
        className="brainhub-image-rotator"
      >
        <img
          className="brainhub-image-rotator__image"
          src="testImage"
          style={
            Object {
              "transform": "rotate(0deg)",
            }
          }
        />
      </div>
      <div
        className="brainhub-image-rotator__controls"
      >
        <button
          className="brainhub-image-rotator__button brainhub-image-rotator__button--arrow"
          onClick={[Function]}
        >
          <svg
            style={
              Object {
                "display": "inline-block",
                "fontSize": "inherit",
                "height": "1em",
                "overflow": "visible",
                "verticalAlign": "-.125em",
                "width": "1em",
              }
            }
            viewBox="0 0 512 512"
          >
            <path
              d="M212.333 224.333H12c-6.627 0-12-5.373-12-12V12C0 5.373 5.373 0 12 0h48c6.627 0 12 5.373 12 12v78.112C117.773 39.279 184.26 7.47 258.175 8.007c136.906.994 246.448 111.623 246.157 248.532C504.041 393.258 393.12 504 256.333 504c-64.089 0-122.496-24.313-166.51-64.215-5.099-4.622-5.334-12.554-.467-17.42l33.967-33.967c4.474-4.474 11.662-4.717 16.401-.525C170.76 415.336 211.58 432 256.333 432c97.268 0 176-78.716 176-176 0-97.267-78.716-176-176-176-58.496 0-110.28 28.476-142.274 72.333h98.274c6.627 0 12 5.373 12 12v48c0 6.627-5.373 12-12 12z"
              fill="currentColor"
            />
          </svg>
        </button>
        <button
          className="brainhub-image-rotator__button brainhub-image-rotator__button--save"
          onClick={[Function]}
        >
          Save image
        </button>
        <button
          className="brainhub-image-rotator__button brainhub-image-rotator__button--arrow"
          onClick={[Function]}
        >
          <svg
            style={
              Object {
                "display": "inline-block",
                "fontSize": "inherit",
                "height": "1em",
                "overflow": "visible",
                "verticalAlign": "-.125em",
                "width": "1em",
              }
            }
            viewBox="0 0 512 512"
          >
            <path
              d="M500.333 0h-47.411c-6.853 0-12.314 5.729-11.986 12.574l3.966 82.759C399.416 41.899 331.672 8 256.001 8 119.34 8 7.899 119.526 8 256.187 8.101 393.068 119.096 504 256 504c63.926 0 122.202-24.187 166.178-63.908 5.113-4.618 5.354-12.561.482-17.433l-33.971-33.971c-4.466-4.466-11.64-4.717-16.38-.543C341.308 415.448 300.606 432 256 432c-97.267 0-176-78.716-176-176 0-97.267 78.716-176 176-176 60.892 0 114.506 30.858 146.099 77.8l-101.525-4.865c-6.845-.328-12.574 5.133-12.574 11.986v47.411c0 6.627 5.373 12 12 12h200.333c6.627 0 12-5.373 12-12V12c0-6.627-5.373-12-12-12z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
    <canvas
      style={
        Object {
          "display": "none",
        }
      }
    />
  </div>
</div>
`;
